tests/dt/test-fillna.py

```diff
@@ -0,0 +1,141 @@
+
+
+def test_fillna_str():
+  assert str(fillna(f.A, reverse=False)) == "FExpr<" + fillna.__name__ + "(f.A, reverse=False)>"
+  assert str(fillna(f.A, reverse=False) + 1) == "FExpr<" + fillna.__name__ + "(f.A, reverse=False) + 1>"
+  assert str(fillna(f.A + f.B, reverse = True)) == "FExpr<" + fillna.__name__ + "(f.A + f.B, reverse=True)>"
+  assert str(fillna(f.B, reverse = True)) == "FExpr<" + fillna.__name__ + "(f.B, reverse=True)>"
+  assert str(fillna(f[:2], reverse = False)) == "FExpr<"+ fillna.__name__ + "(f[:2], reverse=False)>"
+
+
+
+def test_fillna_empty_frame():
+    DT = dt.Frame()
+    expr_fillna = fillna(DT, reverse=False)
+    assert isinstance(expr_fillna, FExpr)
+    assert_equals(DT[:, fillna(f[:], reverse=False)], DT)
+
+
+
+def test_fillna_void():
+    DT = dt.Frame([None, None, None])
+    DT_fillna = DT[:, fillna(f[:], reverse=True)]
+    assert_equals(DT_fillna, DT)
+
+
+
+def test_fillna_trivial():
+    DT = dt.Frame([0]/dt.int64)
+    fillna_fexpr = fillna(f[:], reverse = True)
+    DT_fillna = DT[:, fillna_fexpr]
+    assert isinstance(fillna_fexpr, FExpr)
+    assert_equals(DT, DT_fillna)
+
+
+def test_fillna_bool():
+    DT = dt.Frame([None, False, None, True, False, True])
+    DT_fillna = DT[:, [fillna(f[:], reverse = False), fillna(f[:], reverse = True)]]
+    DT_ref = dt.Frame([
+                [None, False, False, True, False, True],
+                [False, False, True, True, False, True]
+             ])
+    assert_equals(DT_fillna, DT_ref)
+
+
+def test_fillna_small():
+    DT = dt.Frame([None, 3, None, 4])
+    DT_fillna = DT[:, [fillna(f[:], reverse = False), fillna(f[:], reverse = True)]]
+    DT_ref = dt.Frame([
+                  [None, 3, 3, 4],
+                  [3, 3, 4, 4]
+              ])
+    assert_equals(DT_fillna, DT_ref)
+
+
+def test_fillna_string():
+    DT = dt.Frame([None, 'a', None, 'b'])
+    DT_fillna = DT[:, [fillna(f[:], reverse = False), fillna(f[:], reverse = True)]]
+    DT_ref = dt.Frame([
+                  [None, 'a', 'a', 'b'],
+                  ['a', 'a', 'b', 'b']
+              ])
+    assert_equals(DT_fillna, DT_ref)
+
+
+def test_fillna_grouped():
+    DT = dt.Frame([[15, None, 136, 93, 743, None, None, 91], ['a','a','a','b','b','c','c','c']])
+    DT_fillna = DT[:, [fillna(f[:], reverse = False), fillna(f[:], reverse = True)], by(f[-1])]
+    DT_ref = dt.Frame({
+                  'C1':['a','a','a','b','b','c','c','c'],
+                  'C0':[15, 15, 136, 93, 743, None, None, 91],
+                  'C2':[15, 136, 136, 93, 743, 91, 91, 91],
+              })
+    assert_equals(DT_fillna, DT_ref)
+
+
+def test_fillna_grouped_column():
+     DT = dt.Frame([2, 1, None, 1, 2])
+     DT_bfill = DT[:, [fillna(f[0], reverse = False), fillna(f[0], reverse = True)], by(f[0])]
+     DT_ref = dt.Frame([
+                 [None, 1, 1, 2, 2],
+                 [None, 1, 1, 2, 2],
+                 [None, 1, 1, 2, 2]
+              ])
+     assert_equals(DT_bfill, DT_ref)
```

Comment:
When the column is grouped, we could actually have a shortcut: `fillna(col) == col`. An example of how to check if a particular column is grouped: https://github.com/h2oai/datatable/blob/main/src/core/expr/fexpr_qcut.cc#L126-L129

Reply:
Do you mind explaining this more? Is this wrt the tests or the actual function?

Reply:
The actual function. We can detect if the column in question is grouped and then  simply return it as a result, instead of running the complicated filling procedure. That’s because the values are constant within the group of the grouped column.

Reply:
Maybe I am missing the definition of grouped. If you don't mind kindly explain. My notion of grouped is `5, 3,NA,4` belong to group 1. I'm probably misunderstanding the grouped column meaning

Reply:
What you say is a column from the grouped frame, but the frame is grouped by a different column. 

Look at the above example, the content of the column by which the frame is grouped is always a set of groups with constant values.

Reply:
done. makes sense to possibly add it to other pre existing cumulative functions